internal/app/machined/pkg/system/services/registryd.go

```diff
@@ -0,0 +1,479 @@
+		zap.Bool("proxied", isProxied),
+		zap.String("remote_addr", req.RemoteAddr),
+	)
+
+	logger.Info("got request")
+
+	switch req.Method {
+	case http.MethodGet, http.MethodHead:
+		// accepted
+	default:
+		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+
+		return
+	}
+
+	requestPath := path.Clean(req.URL.Path)
+
+	// quickly respond to health check
+	switch requestPath {
+	case "/v2", "/healthz":
+		w.WriteHeader(http.StatusOK)
+
+		return
+	}
+
+	registry := req.URL.Query().Get("ns")
+	if registry == "" {
+		logger.Error("no registry specified")
+		w.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	parts := strings.Split(requestPath, "/")
+	if len(parts) < 5 {
+		logger.Error("wrong paths count")
+		w.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	l := len(parts)
+
+	var (
+		name, digest string
+		isBlob       bool
+	)
+
+	switch {
+	case parts[1] == "v2" && parts[l-2] == "manifests":
+		name = strings.Join(parts[2:l-2], "/")
+		digest = parts[l-1]
+	case parts[1] == "v2" && parts[l-2] == "blobs":
+		name = strings.Join(parts[2:l-2], "/")
+		digest = parts[l-1]
+
+		isBlob = true
+	default:
+		logger.Error("wrong path")
+		w.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	if !reference.NameRegexp.MatchString(name) {
+		w.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	if !reference.DigestRegexp.MatchString(digest) {
+		w.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	logger.Info("image request", zap.String("name", name), zap.String("digest", digest), zap.Bool("is_blob", isBlob), zap.String("registry", registry))
+
+	ref, err := reference.Parse(fmt.Sprintf("%s/%s@%s", registry, name, digest))
+	if err != nil {
+		s.logger.Error("failed to parse reference", zap.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	canonical, ok := ref.(reference.Canonical)
+	if !ok {
+		logger.Error("not a canonical reference")
+		w.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	logger.Info("canonical reference", zap.String("canonical", canonical.String()))
+
+	if isProxied {
+		s.handleLocal(logger.With(zap.String("handler", "local")), w, req, canonical, isBlob)
+	} else {
+		s.handleFanout(logger.With(zap.String("handler", "fanout")), w, req, canonical, isBlob)
+	}
+}
+
+func (s *registrydService) handleLocal(logger *zap.Logger, w http.ResponseWriter, req *http.Request, canonical reference.Canonical, isBlob bool) {
+	var (
+		ctx  context.Context
+		info content.Info
+		err  error
+	)
+
+	for _, namespace := range []string{constants.SystemContainerdNamespace, criconstants.K8sContainerdNamespace} {
+		ctx = namespaces.WithNamespace(req.Context(), namespace)
+
+		info, err = s.client.ContentStore().Info(ctx, canonical.Digest())
+		if err != nil {
+			if errdefs.IsNotFound(err) {
+				continue
+			}
+
+			logger.Error("failed to get content info", zap.Error(err))
+			w.WriteHeader(http.StatusInternalServerError)
+
+			return
+		}
+
+		break
+	}
+
+	if err != nil {
```

Comment:
this seems not needed, which err case is this handling?

Reply:
this handles the case when we get `IsNotFound` for all namespaces we checked for